src/main/scala/com/evolutiongaming/crypto/Crypto.scala

```diff
@@ -1,149 +1,226 @@
+      new String(cipher.doFinal(encryptedData), UTF_8)
+    }
   }
 
   /**
-    * Generates the SecretKeySpec, given the private key and the algorithm.
+    * Current AES mode - V3:
+    * - no restrictions on key size - SHA256 hash is used to obtain key entropy
+    * - AES/GCM/NoPadding (128 bit key) cipher is used to provide authenticated encryption
+    * - dynamic length random IV - 12 bytes by default with possible extension up to 267 bytes
+    * - 128 bit auth tag length
     */
-  private def secretKeyWithSha256(privateKey: String, algorithm: String) = {
-    val messageDigest = MessageDigest.getInstance("SHA-256")
-    messageDigest.update(privateKey.getBytes("utf-8"))
-    // max allowed length in bits / (8 bits to a byte)
-    // For AES we hardcode keylength to 128bit minimum to not depend on environment security policy settings:
-    // it may vary between 128 and 256 bits which can yield different encryption keys if we don't
-    val maxAllowedKeyLength = if (algorithm == "AES") AesKeyBytesMaxSize else Cipher.getMaxAllowedKeyLength(algorithm) / 8
-    val raw = messageDigest.digest().slice(0, maxAllowedKeyLength)
-    new SecretKeySpec(raw, algorithm)
+  private object AES_V3 {
+    /*
+    implementation based on
+    https://proandroiddev.com/security-best-practices-symmetric-encryption-with-aes-in-java-7616beaaade9
+     */
+
+    private val CipherTransformation = "AES/GCM/NoPadding"
+    /*
+    same auth tag length as in
+    https://proandroiddev.com/security-best-practices-symmetric-encryption-with-aes-in-java-7616beaaade9
+     */
+    private val AuthTagLengthBits = 128
+    /*
+    for GCM IV a 12 byte random byte-array is recommend by NIST
+    https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf
+     */
+    private val CurrentIVLengthBytes = 12
+    private val MinIVLengthBytes = 12 //does not allow decrypting with IVs smaller than 12 bytes (96 bits)
+
+    def encrypt(value: String, privateKey: String): String = {
+      val skeySpec = aesSKey128bitWithSha256(privateKey.getBytes(UTF_8))
+      val iv = new Array[Byte](CurrentIVLengthBytes)
+      secureRandom.nextBytes(iv)
+      val parameterSpec = new GCMParameterSpec(AuthTagLengthBits, iv)
+      val cipher = Cipher.getInstance(CipherTransformation)
+      cipher.init(Cipher.ENCRYPT_MODE, skeySpec, parameterSpec)
+      val encryptedValue = cipher.doFinal(value.getBytes(UTF_8))
+      encodeEncryptedToString(iv, encryptedValue)
+    }
+
+    private def encodeEncryptedToString(iv: Array[Byte], encryptedValue: Array[Byte]): String = {
+      //1 byte for dynamic IV length encoding
+      val buf = ByteBuffer.allocate(1 + iv.length + encryptedValue.length)
+      require(iv.length >= MinIVLengthBytes, s"IV length shouldn't be smaller than $MinIVLengthBytes bytes")
+      //encode IV length as an unsigned byte starting from MinIVLengthBytes
+      buf.put((iv.length - MinIVLengthBytes).toByte)
```

Comment:
Doesn't look like a good idea to me. At some point you might want to bump the minimal threshold and it will break all the data encrypted before. What is the benefit of adding this complexity instead of just validating input of this method? It makes this code less portable. My opinion that encoded value should contain all the data needed for decryption and not rely on any magic numbers.

Reply:
Even having 12 of all available options useless still leaves a lot of room in range 0-255, what should be totally enough for foreseeable future.

Reply:
There won't be a need to bump threshold, but other than that, I agree with @arixmkii. No need for such complexity. Usable range 12-255 will still be fine.

Reply:
Agree. I've changed length encoding as you suggested, please check